src/modules/iotjs_module_mqtt.c

```diff
@@ -0,0 +1,719 @@
+/* Copyright 2018-present Samsung Electronics Co., Ltd. and other contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "iotjs_def.h"
+#include "iotjs_module_buffer.h"
+#include "iotjs_module_mqtt.h"
+
+
+#include "iotjs_handlewrap.h"
+#include "iotjs_reqwrap.h"
+
+
+static jerry_value_t mqtt_client_connack_error(const unsigned char error_code) {
+  switch (error_code) {
+    case UNACCEPTABLE_PROTOCOL:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: unacceptable protocol");
+    case BAD_IDENTIFIER:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: bad client identifier");
+    case SERVER_UNAVIABLE:
+      return JS_CREATE_ERROR(COMMON,
+                             "MQTT: Connection refused: server unaviable");
+    case BAD_CREDENTIALS:
+      return JS_CREATE_ERROR(
+          COMMON, "MQTT: Connection refused: bad username or password");
+    case UNAUTHORISED:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Connection refused: unauthorised");
+    default:
+      return JS_CREATE_ERROR(COMMON, "MQTT: Unknown error");
+  }
+}
+
+
+static uint8_t *iotjs_encode_remaining_length(uint32_t len,
+                                              unsigned char *buffer) {
+  size_t rc = 0;
+  do {
+    unsigned char d = len & 0x7F;
+    len >>= 7;
+    if (len > 0) {
+      d |= 0x80;
+    }
+    buffer[rc++] = d;
+  } while (len > 0);
+
+  return (buffer + rc);
+}
+
+
+static size_t get_remaining_length_size(uint32_t len) {
+  uint8_t n = 0;
+  while (len != 0) {
+    len >>= 7;
+    n++;
+  }
+
+  return n;
+}
+
+
+static uint32_t iotjs_decode_remaining_length(char *buffer, size_t *offset) {
+  unsigned char c;
+  uint32_t remaining_length = 0;
+  uint32_t length = 0;
+  uint32_t multiplier = 1;
+
+  do {
+    if (++length > IOTJS_MODULE_MQTT_MAX_REMAINING_LENGTH_BYTES) {
+      return UINT32_MAX;
+    }
+    c = (unsigned char)buffer[*offset];
+    remaining_length += (c & 127) * multiplier;
+    multiplier *= 128;
```

Comment:
`multiplier <<= 7`

Reply:
Btw usually shift is faster than multiplication, so you could simply use a shift:
```
remaining_length += (c & 0x7F) << shift;
shift += 7
```
Also `|=` can be used instead of `+=` here, but they are equally fast so they don't provide any advantage.